compute_sdk/globus_compute_sdk/sdk/bash_function.py

```diff
@@ -0,0 +1,249 @@
+             with kwargs passed at call time.
+
+        stdout: str | None
+            file path to which stdout should be captured
+
+        stderr: str | None
+            file path to which stderr should be captures
+
+        walltime: float | None
+            duration in seconds after which the command should be interrupted
+
+        rundir: str | None
+            directory within which the command should be executed
+
+        run_in_sandbox: bool
+            when set, the command will execute within a directory matching
+            the task UUID
+            default=True
+
+        snippet_lines: int
+            Number of lines of stdout/err to capture,
+            default=1000
+
+        """
+        self.cmd = cmd
+        self.stdout = stdout
+        self.stderr = stderr
+        self.walltime = walltime
+        self.rundir = rundir
+        self.run_in_sandbox = run_in_sandbox
+        self.snippet_lines = snippet_lines
+
+    @property
+    def __name__(self):
+        # This is required for function registration
+        return self.cmd
+
+    def open_std_fd(self, fname, mode: str = "a+"):
+        import os
+
+        # fname is 'stdout' or 'stderr'
+        if fname is None:
+            return None
+
+        if os.path.dirname(fname):
+            os.makedirs(os.path.dirname(fname), exist_ok=True)
+        fd = open(fname, mode)
+        return fd
+
+    def get_snippet(self, file_obj) -> str:
+        file_obj.seek(0, 0)
+        last_n_lines = file_obj.readlines()[-self.snippet_lines :]
+        return "".join(last_n_lines)
+
+    def get_and_close_streams(self, stdout, stderr) -> t.Tuple[str, str]:
+        stdout_snippet = self.get_snippet(stdout)
+        stderr_snippet = self.get_snippet(stderr)
+        stdout.close()
+        stderr.close()
+        return stdout_snippet, stderr_snippet
+
+    def execute_cmd_line(
+        self,
+        cmd: str,
+        stdout: t.Optional[str] = None,
+        stderr: t.Optional[str] = None,
+        rundir: t.Optional[str] = None,
+    ) -> BashResult:
+        import os
+        import subprocess
+
+        run_dir = rundir or self.rundir
+
+        if not run_dir:
+            # run_dir takes priority over sandboxing
+            if os.environ.get("GC_TASK_SANDBOX_DIR"):
+                run_dir = os.environ["GC_TASK_SANDBOX_DIR"]
+            else:
+                if self.run_in_sandbox and os.environ.get("GC_TASK_UUID"):
+                    run_dir = os.environ["GC_TASK_UUID"]
+                    os.environ["GC_TASK_SANDBOX_DIR"] = os.path.join(
+                        os.getcwd(), run_dir
+                    )
+
+        if run_dir:
+            os.makedirs(run_dir, exist_ok=True)
+            os.chdir(run_dir)
+
+        stdout = (
+            stdout or self.stdout or tempfile.NamedTemporaryFile(dir=os.getcwd()).name
+        )
+        stderr = (
+            stderr or self.stdout or tempfile.NamedTemporaryFile(dir=os.getcwd()).name
+        )
```

Comment:
When running bash functions locally, these file names can be a little hard to track. Discussed some options on a call:

* add `.stderr`/`.stdout` extensions
* put a counter in the filename for context on run ordering
* include `__name__()` in filename?

This is a bit of a larger thought process, so it doesn't need to be in this PR

Reply:
I definitely agree with adding the `.stdout` and `.stderr` file extensions. We could also consider using the task UUID for the filename.

Reply:
I like these ideas 👍 
I've added the stdout/err extensions as tempfile suffixes.

I'd keep more identifiable names for later. I'd have liked to put the task id in the name, but that's going to be in the task working dir if that were available on the endpoint, so I think that'd be redundant.

Reply:
But the task working dir will only be created if `run_in_sandbox` is set to true, right?

Reply:
From discussion:
1. We agreed to have the prefix get set to the GC_TASK_UUID field.
2. If the GC_TASK_UUID is not set, say shellfunction is sent to an older endpoint, we set the prefix to a random string, that will at least group the stdout/err files from the same run.

Reply:
This is now fixed.